### DISCUSSION TITLE
Request for information on LivingDoc support

### DISCUSSION THREAD

**Discussion by ptsumbe** [reqnroll/Reqnroll#68] [Q&A] [answered]
Hi @gasparnagy, Really delighted to see new offering started by you. Currently we use SpecFlow with LivingDoc. What we have seen so far Reqnroll does not support it, can you please help comment when will it be available ? Do we have any prospect dates in mind ? #LivingDoc support

**Comment by gasparnagy** [comment, marked as answer]
Thank you @ptsumbe for your support. 

I think the problem you facing is a common problem for many of us, so solving it should be a priority problem. (Especially because the [SpecFlow.Plus.LivingDoc.CLI](https://www.nuget.org/packages/SpecFlow.Plus.LivingDoc.CLI) supports only .NET 3.1 and .NET 5, both are out of support already by Microsoft.)

We are currently busy with stabilizing the code base, but the next thing after should be the support for living doc. We track the efforts at #37. As we all work on this on a voluntary basis I think it would not be fair to set any deadline, but I hope to have something available before summer. Also our I got the information from the Allure team that they are also working on their Reqnroll integration that will be ready in the next weeks, so Allure reports will become a good alternative / workaround until the living doc support is implemented.

**Reply by ptsumbe** [reply]
Thank you @gasparnagy for response. It really helped!

**Reply by fkurzyk** [reply]
Thanks you @gasparnagy for this answer and for all your current and past efforts on .NET BDD.
I would like to extend ptsumbe's question about LivingDoc support and ask specifically about planned scope of LivingDoc replacement. Do you plan to have an Azure DevOps integration? Currently it is possible to view LivingDoc natively in AzDO, it would be great is it was also possible with Reqnroll's version.

**Reply by gasparnagy** [reply]
@fkurzyk The short answer would be that making the LivingDoc integrated to Azure DevOps is not fully out of scope, but obviously we first need the solution to generate HTML from the feature files & result. After that, we should re-evaluate if/how we could host that on Azure DevOps.

**Comment by ThomasHeijtink** [comment]
Just to mention it here as well. You can easily create a `TestExecution.json` file yourself in order to make the LivingDocs work again. So this doesn't have to be a deal breaker anymore to switch to Reqnroll.

**Reply by gasparnagy** [reply]
@ThomasHeijtink Could you share a bit more info about that? What does you build script uses as an input to generate the TestExecution.json?

**Reply by ThomasHeijtink** [reply]
@gasparnagy I hope this is what you are after.

I've created a MSBuild task which splits up all our feature files for the ultimate scenario level parallelisation. In that code I loop through all the feature files and its scenarios, and in case of a scenario outline I also make a separate feature file per example. Coincidentally, this is also the level of input required by the LivingDocs.

I have a `.props` file like this

````XML
<Project TreatAsLocalProperty="TaskFolder;TaskAssembly">
    <PropertyGroup>
        <CustomTasksFolder>..\tasks\net8.0</CustomTasksFolder>
        <CustomTasksAssembly>$(CustomTasksFolder)\Common.Interface.Testing.MSBuild.dll</CustomTasksAssembly>
    </PropertyGroup>

    <UsingTask TaskName="Common.Interface.Testing.MSBuild.SplitFeatureFiles" AssemblyFile="$(CustomTasksAssembly)"/>

    <PropertyGroup>
        <RootFolder Condition="'$(RootFolder)' == ''">$(MSBuildProjectDirectory)</RootFolder>
    </PropertyGroup>
</Project>
````
A `.targets` file like this
```XML
<Project>
    <Import Project="Common.Interface.Testing.MSBuild.props" />
        
    <ItemGroup>
        <FeatureFiles Include="$(RootFolder)\Features\*.feature" />
    </ItemGroup>

    <Target Name="SplitFeatureFiles" BeforeTargets="UpdateFeatureFilesInProject" DependsOnTargets="BeforeUpdateFeatureFilesInProject">
        <ItemGroup>
            <ReqnrollFeatureFiles Remove="SplitFeatures\*.feature"/>
            <Compile Remove="SplitFeatures\*.feature.cs"/>
        </ItemGroup>
        <SplitFeatureFiles FeatureFiles="@(FeatureFiles)" OutputPath="$(OutputPath)" />
        <ItemGroup>
            <ReqnrollFeatureFiles Include="SplitFeatures\*.feature"/>
            <Compile Include="SplitFeatures\*.feature.cs"/>
        </ItemGroup>
    </Target>
</Project>
```

My task looks something like this:
```CSharp
public class SplitFeatureFiles : Task {

    private readonly string _newLine = Environment.NewLine;
    private readonly string _directory = "SplitFeatures";
    private readonly List<string> _splitFeatureFileNames = new();

    public string OutputPath { get; set; } = string.Empty;
    
    [Required]
    public ITaskItem[] FeatureFiles { get; set; }
    
    public override bool Execute() {

        var testResult = new TestExecution(DateTime.Now);
        
        if (Directory.Exists($"./{_directory}") == false)
            Directory.CreateDirectory($"./{_directory}");
        
        foreach (var featureFile in FeatureFiles) {
            SplitFeatureFile(featureFile.GetMetadata("FullPath"), testResult);
        }

        var filesToDelete = Directory.GetFiles($"./{_directory}", "*.feature")
                                     .Where(f => _splitFeatureFileNames.Contains(Path.GetFileName(f)) == false)
                                     .ToArray();
        
        foreach (var fileToDelete in filesToDelete) {
            try {
                File.Delete(fileToDelete);
                File.Delete($"{fileToDelete}.cs");
            }
            catch (Exception e) when (e is IOException or UnauthorizedAccessException) {
                Log.LogErrorFromException(e, showStackTrace: true);
            }
            catch {
                //Probably already gone
            }
        }
        
        var serializerOptions = new JsonSerializerOptions
        {
            Encoder       = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
            WriteIndented = true
        };

        if (string.IsNullOrEmpty(OutputPath) == false && 
            Directory.Exists(OutputPath) == false) {
            Directory.CreateDirectory(OutputPath);
        }
        
        File.WriteAllText($"{OutputPath}{nameof(TestExecution)}.json", JsonSerializer.Serialize(testResult, serializerOptions), Encoding.UTF8);
        
        return Log.HasLoggedErrors == false;
    }

    private void SplitFeatureFile(string filePath, TestExecution testExecution) {

        var lines = File.ReadAllLines(filePath);
        
        var splitSettings = lines.Any(l => l.Contains("language: en"))
            ? SplitSettings.GetEnglish()
            : SplitSettings.GetDutch();

        var featureTitle = lines.FirstOrDefault(x => x.Contains(splitSettings.FeatureKeyword))?.Trim();

        if (featureTitle == null) {
            return;
        }
        
        var featureTitleWithTags = featureTitle;
        featureTitle = string.Join(':', featureTitle.Split(':', StringSplitOptions.TrimEntries).Skip(1));
        
        var content = string.Join(
            _newLine, 
            lines.Select(l => l.Trim()
                               .Replace(splitSettings.ScenarioOutlineKeyword, splitSettings.ScenarioOutlineMockKeyword))
                 .Where(l => l.StartsWith('#') == false));

        var sectionsAndTags = Regex.Split(content, splitSettings.SplitRegexExpression)
                                   .Where(s => string.IsNullOrEmpty(s) == false)
                                   .Select(s => s.Replace(splitSettings.ScenarioOutlineMockKeyword,
                                       splitSettings.ScenarioOutlineKeyword)).ToArray();
        
        var tags       = new HashSet<string>(){$"@split{_newLine}"};
        var background = string.Empty;
        var scenarios  = new List<string>();
        foreach (var sectionOrTag in sectionsAndTags) {
            if (sectionOrTag.StartsWith('@')) {
                tags.Add(sectionOrTag);
                continue;
            }

            if (sectionOrTag.StartsWith(splitSettings.FeatureKeyword)) {
                featureTitleWithTags = tags.Count > 0 ? $"{string.Join(_newLine, tags)}{featureTitleWithTags}" : featureTitleWithTags ;
            }
            else if (sectionOrTag.StartsWith(splitSettings.BackgroundKeyword)) {
                background = $"{sectionOrTag}";
            }
            else {
                scenarios.Add($"{string.Join(_newLine, tags)}{sectionOrTag}");
            }
            
            tags.Clear();
        }

        var scenarioCount = 1;
        foreach (var scenario in scenarios) {
            var examples        = new []{""};
            var scenarioSection = scenario;
            var exampleHeaders  = string.Empty;
            var scenarioTitle   = string.Join(':', scenario.Split(_newLine)[0].Trim().Split(':').Skip(1)).Trim();
            if (scenario.StartsWith(splitSettings.ScenarioOutlineKeyword)) {
                var scenarioSections = scenario.Split(splitSettings.ExampleKeyword, StringSplitOptions.TrimEntries);
                scenarioSection  = scenarioSections.First();
                var exampleSection   = scenarioSections.Last();
                examples = exampleSection.Split(_newLine, StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries); 
                exampleHeaders = $"{splitSettings.ExampleKeyword}{_newLine}{examples[0]}";
                examples = examples.Skip(1).ToArray();
            }

            foreach (var (count, example) in examples.Select((e, i) => (i+1, e))) {

                var numberedScenario = scenarioSection.Replace(
                    $"{splitSettings.ScenarioOutlineKeyword} ",
                    $"{splitSettings.ScenarioOutlineKeyword} {count}_");

                var featureContent = $"""
                                      {splitSettings.LanguageIndicator}{featureTitleWithTags} {scenarioCount++}

                                      {background}{numberedScenario}
                                      """;

                if (example != string.Empty) {
                    featureContent += $"{_newLine}{_newLine}{exampleHeaders}{_newLine}{example}";
                }

                var newFileName = Path.GetFileNameWithoutExtension(filePath) + $"{scenarioCount - 1}.feature";

                File.WriteAllText($"./{_directory}/{newFileName}", featureContent);

                var stepArguments = example.Split('|', StringSplitOptions.TrimEntries);
                stepArguments = stepArguments.Skip(1).Take(stepArguments.Length - 2).ToArray();
                
                _splitFeatureFileNames.Add(newFileName);
                var executionResult = new ExecutionResult(
                    featureTitle, 
                    scenarioTitle,
                    stepArguments);
                
                testExecution.ExecutionResults.Add(executionResult);
            }
        }
    }
}

````

I use this parent class:
```CSharp
internal class TestExecution(DateTime executionTime) {
    public string[] Nodes => [];
    public DateTime ExecutionTime => executionTime;
    public DateTime GenerationTime { get; }
    public Guid PluginUserSpecFlowId => new Guid("36bde1dc-fd2c-45e7-a4f2-fae8799bfad4"); //This is some sort of Guid which I could find on our build server. It was always the same in all test runs but differed from the Guid value that was used on my local machine.
    public string? CLIUserSpecFlowId { get; }

    public List<ExecutionResult> ExecutionResults { get; } = new();

    public string? StepReports { get; }
}
```
And each scenario and example represent an ExecutionResult like this:
```CSharp
internal class ExecutionResult(
    string featureTitle,
    string scenarioTitle,
    string[] stepArguments) {
    public string ContextType => "Scenario";
    public string FeatureFolderPath => "Features";

    public string FeatureTitle => featureTitle;
    public string ScenarioTitle => scenarioTitle;
    public string[] StepArguments => stepArguments;
    
    public string Status => "OK";
    public string[] StepResults => [];
    public string? Outputs { get; }
}
```

When you write this to a `TestExecution.json` file in a directory depending on your configuration of your `SpecFlowPlus@0` task. It's normally generated in the OutputDir for the build.

It will be picked up and used for the LivingDoc.

This will result in something like this:
```JSON
{
  "Nodes": [],
  "ExecutionTime": "2024-04-23T14:37:07.1205458+02:00",
  "GenerationTime": "0001-01-01T00:00:00",
  "PluginUserSpecFlowId": "36bde1dc-fd2c-45e7-a4f2-fae8799bfad4",
  "CLIUserSpecFlowId": null,
  "ExecutionResults": [
    {
      "ContextType": "Scenario",
      "FeatureFolderPath": "Features",
      "FeatureTitle": "Example feature",
      "ScenarioTitle": "Example scenario",
      "StepArguments": [],
      "Status": "OK",
      "StepResults": [],
      "Outputs": null
    },
    {
      "ContextType": "Scenario",
      "FeatureFolderPath": "Features",
      "FeatureTitle": "Example feature",
      "ScenarioTitle": "Example scenario outline",
      "StepArguments": [
        "example value 1",
        "example value 2",
        "example value 3",
        "example value 4",
        "example value 5"
      ],
      "Status": "OK",
      "StepResults": [],
      "Outputs": null
    }
  ],
  "StepReports": null
}
```

I our case we weren't interested in the actual test results as this is part of our pipeline. So I didn't care about it that much. Only the regular scenarios, as opposed the scenario outlines, were marked as passed. Even though I reported everything was OK. So maybe the scenario outlines require something extra which I haven't found yet. Maybe they require at least a single step result. But in the StepResults you can add objects which represent these values.

The step results look like this:
```JSON
{"Duration":"00:00:00.0053099","Status":"OK","Error":null,"Outputs":null}
```

I hope it helps someone to make the step to Reqnroll!

**Comment by adae-palfinger** [comment]
Any update on this?

**Reply by gasparnagy** [reply]
I think most of us are on vacation (or at least me), so real progress can only be expected in September.

**Reply by pritesh848** [reply]
Any update on this?

**Comment by dakiers** [comment]
So we actually just recently started a SpecFlow project and realized it wasn't supported anymore. Converted to Reqnroll and the only thing missing for us was the LivingDoc which we're using the DevOp plugin for.
I created some Hooks that creates the "TestExecution.json" that is published. For now, that's enough for us.

I extracted the code [here](https://github.com/dakiers/reqnroll/tree/main/LivingDoc)

Feel free to use the code in anyway you feel fit.

**Reply by mouradchama** [reply]
Hi thanks for the code, did you forget to share the class OutputHelper? I'm not able to compile your code, this is the only missing reference.

**Reply by dakiers** [reply]
Yeah suppose I did. It's something I was playing around with. Suppose it should be IReqnrollOutputHelper.

Anyway I added OutputHelper to the repository.

**Reply by quicoli** [reply]
Hi @dakiers sorry for the silly question... but how do I use your code?

**Reply by dakiers** [reply]
@quicoli you might also want to take a look at https://github.com/orgs/reqnroll/discussions/355

I extracted the code from project I am working for. Since I'm lazy, I didn't bother to put it in a solution or anything.

Anyway you should be able to just copy the code to your project, probably want to give everything a nice namespace, and it should just work from there.
If you run test automations locally, the "TestExecution.json" should be placed in your "bin" folder.

Oh I did update the repository with the Error fix.

**Reply by quicoli** [reply]
Thank you! I started doing what you suggested and so far I can see the "TestExecution.json" being created. I also implemented the the error fix here.  

Your files could be integrated in the main project and be distributed within the reqnroll package, couldn't? 
It's a great solution and would help many people I think.  

Anyway, thanks a million! 👍🏻

**Reply by dakiers** [reply]
@quicoli In theory you could just make it a [Plugin](https://docs.reqnroll.net/latest/extend/plugins.html) in a separate class library project. Just having the project referenced should be enough. Just like it was for SpecFlow.LivingDoc.

Can't say if the code works for everyone (for every situation) though. At the very least it can help them on their way.

In our case the code is part of a larger library (for internal use) we're using as a base for all test automation projects. It's NuGet package which its assembly is added to the reqnroll.json's bindingAssemblies (to bind the Hooks and StepDefinitions of the library).
We might change it to be a plugin instead, but this works fine for now.

**Reply by saikabade** [reply]
Hi @quicoli  I was able to use your code suggestions to generate the TestExecution.json file and also able to generate the livingDoc html with the BDD documentation. However, it was not able to generate the actual test execution test results like the Passed and Failed tests, and also was not able to show the logs that are created using the WriteLine(string message) of IReqnrollOutputHelper. It just displays the plain features. Can you please suggest?

**Reply by Registeel** [reply]
@saikabade Just confirming. 
1. Your Test Execution JSON path is for sure pointing to the generated TestExecution.json file with your test results in it?
2. You have the "Test results" switch in the SpecFlow+ LivingDoc html page turned on?

**Reply by saikabade** [reply]
1. Yes, am able to see the features in the html
2. yes, the Test results slider switch is on

**Reply by RenoG1** [reply]
Hello,

When using AzureDevops plugin, TestExecution.json well generated, the error message when generating Living Doc is : Warning: No bindings found in assembly 'xxx\s\BDDTests\bin\Release\net8.0\BDDTests.dll'
And the report only shows Scenarii, Steps etc... but status of each test is 'others' (in grey) rather than 'passed' or 'failed'

**Reply by 304NotModified** [reply]
> I extracted the code [here](https://github.com/dakiers/reqnroll/tree/main/LivingDoc)

What is the source of the exacted code? Is this private code, this could be illegal. See also https://github.com/orgs/reqnroll/discussions/68#discussioncomment-12364527

**Comment by gasparnagy** [comment]
See notes of @Registeel at https://github.com/orgs/reqnroll/discussions/355

**Comment by geertjanthomas** [comment]
Hi,
We ran into this issue as well. Moving from Specflow to Reqnroll was fairly straightforward apart from the missing LivingDoc functionality.

In the end I have reverse engineered the Specflow/LivingDoc assemblies using dotPeek and created my own implementation of the plugin and the CLI with minor changes to the reverse engineered code and using as much as possibly already existing Reqnroll functionality.

I'm happy to share the code for those interested.

**Reply by RajeshPoola** [reply]
@geertjanthomas Please share it

**Reply by geertjanthomas** [reply]
I'm working on a github repo and NuGet package. Will be .net48 for now

**Reply by gasparnagy** [reply]
While the motivation is clear, I have to highlight that reverse engineering and recompiling SpecFlow LivingDoc is illegal and breaking the copyright law, as we don't have permission to use those assets.

With Reqnroll we are committed to provide a legal alternative for SpecFlow, so we will not be able to include or promote this solution officially. 

Please everyone keep this in mind and decide responsibly.

**Reply by 304NotModified** [reply]
I'm afraid @gasparnagy is correct, this is illegal. 

I would advise to @geertjanthomas to contact Tricentis and get an approval. 

Also I think we should stop discussing and placing this under the Reqnroll name.

**Reply by geertjanthomas** [reply]
Hi @gasparnagy and @304NotModified ,

Thank you for your replies. I suspected as much about the legality. I will not be publishing anything directly derived from re-engineered code. My intentions are sincere.

I will be looking into options to provide similar functionality to my stakeholders, inspired by earlier implementations (did I phrase that carefully enough 😉).

I'm happy to discuss any results in a more private setting to see if it can benefit the Reqnroll community.

Cheers.

**Reply by gasparnagy** [reply]
@geertjanthomas yes, definitely. While on the surface the things seem to be silent, we already made progress via the cucumber messages PR of @clrudolphi. I was even able to generate HTML with that improvement (https://www.linkedin.com/posts/gasparnagy_cucumber-renqroll-collaboration-activity-7294732223524585472-oPME?utm_source=share&utm_medium=member_desktop&rcm=ACoAAATpLNEBBCcBlgg18lBukrre8ibE3nrtC_k) but we haven't fully finished it yet.

**Comment by quicoli** [comment]
> Hi,
> We ran into this issue as well. Moving from Specflow to Reqnroll was fairly straightforward apart from the missing LivingDoc functionality.
> 
> In the end I have reverse engineered the Specflow/LivingDoc assemblies using dotPeek and created my own implementation of the plugin and the CLI with minor changes to the reverse engineered code and using as much as possibly already existing Reqnroll functionality.
> 
> I'm happy to share the code for those interested.


Please share it

**Reply by zdmta** [reply]
i was able to setup allure report together with Azure devops plugin, which looks promising

**Reply by 304NotModified** [reply]
As discussed above, see https://github.com/orgs/reqnroll/discussions/68#discussioncomment-12364527, this is illegal.

I think we should stop discussing and placing this under the Reqnroll name.

**Comment by MarichkaOK** [comment]
There are several workarounds and alternative solutions (but it seems they have problems according to feedbacks) here and I understand that it's illigal to "copy" solution from SpecFlow.Plus.LivingDocPlugin , but my main question is if you are planning to make some reporting solution for Reqnroll? May be similar to what we had with that plugin? What is your official plans about it?
We have being working with SpecFlow for many years and built around this plugin a reporting system. So it updates LivingDoc page on AzureDevops, we implemented solution that accumulates tests results from several runs and report them to AzureDevops. Also we send current tests report in internal company message system. So, when it turned out that SpecFlow project is ended up, these problems with test reporting is a stumbling block for our migration. What is the official proposal about that from the side of Reqnroll?

**Comment by gasparnagy** [comment]
FYI: https://reqnroll.net/news/2025/08/reqnroll-v3-0-released-with-reporting/

**Comment by pritesh848** [comment]
Thanks for this functionality and more interested in T unit features it
looks more qa friendly (similar like testing)

Regards,
Pritesh Muli
